install/TOOLS/dockapps/wmudmount/wmudmount.1
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.TH wmudmount 1 "December 16, 2018"
.SH NAME
wmudmount \- A WindowMaker filesystem mounting dockapp using udisks
.P
.SH SYNOPSIS
.B wmdgumount
.RB [\| options \|]
.SH DESCRIPTION
\fBwmudmount\fP is a filesystem mounter that uses udisks to handle
notification of new volumes and mounting of the filesystems as a non-root
user. It also includes a mode to display the mounted filesystems with the least
free space percentage (similar to wmfsm).
.SH OPTIONS
.TP
.B --help
Display usage information.
.TP
.BR -V ", " --version
Display the version number and exit.
.TP
.BI --no-default-config
Prevent reading of the default configuration files.
.TP
.BI --conf " <filename>"
Specify a configuration file to use in addition to the default.
.TP
.BR -v ", " --verbose
Print more messages to the console. May be repeated.
.TP
.BR -q ", " --quiet
Print fewer messages to the console. May be repeated.
.TP
.BI "-n\fR,\fP --notify" " <level>"
Specify what level of notification to display (using the desktop notification service). Valid levels are:

none - No messages
.br
error - Only provide failure messages, e.g. "Mount failed"
.br
warn - Only provide failure and warning messages
.br
info - Provide informational messages, e.g. "Device added" (default)
.TP
.BR --exclude-system-volumes ", " --include-system-volumes
.PD 0
.TP
.BR --fsm-exclude-system-volumes ", " --fsm-include-system-volumes
.TP
.BR --pager-exclude-system-volumes ", " --pager-include-system-volumes
.PD
Whether to include volumes in the paging list and/or the FSM display that require the administrator authorization for mounting (i.e. those reported as "system internal" by udisks and those listed in fstab without the "user" option).
.TP
.BI --return-to-fsm-timeout " <seconds>"
Specify the number of seconds the mouse pointer must be away from the dockapp before the display changes back to the FSM view. The default is 60, set negative to disable.
.TP
.BR --allow-insecure-memory ", " --no-allow-insecure-memory
Normally, the password dialog will insist that secure memory be allocated to hold the entered password; this can prevent the password from being written to swap in a low-memory situation. If your system does not allow users to allocate secure memory, you can use \fB--allow-insecure-memory\fP to allow the use of normal, insecure memory.
.TP
.BR --allow-core-files ", " --no-allow-core-files
The dumping of core files is disabled so entered passwords will not be written to disk if a coredump signal is received. Normally, the program will exit if this cannot be done; if necessary, \fB--allow-core-files\fP may be used to allow execution despite the risk.
.TP
.BR --blink-full-filesystems ", " --no-blink-full-filesystems
Normally, the "disk space used" bar for a read-write filesystem will blink when it is over 95% full. \fB--no-blink-full-filesystems\fP may be used to prevent this.
.TP
.B --non-wmaker
Create windows differently, in a manner more friendly to non-Window Maker window managers. If you wind up with both a docked icon and an unmanaged window, try setting this.
.TP
.BR --use-secret-service ", " --no-secret-service
Whether to use the Secrets service for encrypted filesystem unlock passphrase storage.
.SH CONFIGURATION
The configuration files may contain the following options:
.TP
.BI verbose
.PD 0
.TP
.BI quiet 
.TP
.BI non-wmaker
.TP
.BI notify "\fR = \fP<level>"
.TP
.BI exclude-system-volumes
.TP
.BI include-system-volumes
.TP
.BI fsm-exclude-system-volumes
.TP
.BI fsm-include-system-volumes
.TP
.BI pager-exclude-system-volumes
.TP
.BI pager-include-system-volumes
.TP
.BI return-to-fsm-timeout "\fR = \fP<seconds>"
.TP
.BI allow-insecure-memory
.TP
.BI no-allow-insecure-memory
.TP
.BI allow-core-files
.TP
.BI no-allow-core-files
.TP
.BI blink-full-filesystems
.TP
.BI no-blink-full-filesystems
.TP
.BI use-secret-service
.TP
.BI no-secret-service
.PD
Same as the corresponding command line options.
.TP
.BI warn "\fR = \fP<level>"
Sets the verbosity level, \fIerror\fP, \fIwarn\fP (default), \fIinfo\fP, or \fIdebug\fP.
.TP
.BI command "\fR = \fP<command-spec>"
Allows adding arbitrary commands to the right-click menu. A command-spec is of
the form \fItitle\fP;\fIcommand\fP. The command is NOT passed through the shell, but the search path is used. The following printf-style escapes may be used:

\fB%p\fP - The udisks dbus path for the device
.br
\fB%d\fP - The /dev path for the device
.br
\fB%m\fP - The mountpoint of the filesystem
.br
\fB%%\fP - A literal "%"

If any of the above are not available (e.g. the filesystem is not mounted), commands using them will be greyed-out in the right-click menu.

For example, "command = Console; urxvt \-cd %m" would create a menu item that will open a terminal window at the filesystem's mountpoint.
.TP
.BI click-command "\fR = \fP<command-spec>"
Same as \fIcommand\fP, plus this command will be executed when clicking a device in the display.
.TP
.BI double-click-command "\fR = \fP<command-spec>"
Same as \fIcommand\fP, plus this command will be executed when double-clicking a device in the display.
.TP
.BI include-volume "\fR = \fP<spec>"
.PD 0
.TP
.BI exclude-volume "\fR = \fP<spec>"
.TP
.BI fsm-include-volume "\fR = \fP<spec>"
.TP
.BI fsm-exclude-volume "\fR = \fP<spec>"
.TP
.BI pager-include-volume "\fR = \fP<spec>"
.TP
.BI pager-exclude-volume "\fR = \fP<spec>"
.PD
Allows for specificly identifying volumes to include or exclude from the paging and/or the FSM display, overriding the general behavior chosen by \-\-include\-system\-volumes or \-\-exclude\-system\-volumes.

The values for \fI<spec>\fP are similar to those for the first field in /etc/fstab: If it begins with "LABEL=", the remainder of the spec is matched against the volume's label. If it begins with "UUID=", the remainder of the spec is matched against the volume's UUID. Otherwise, the spec is matched against the volume's device and current mountpoint.

If multiple include and exclude entries match a volume, the last one wins. Specifying include\-volume is equivalent to specifying both fsm\-include\-volume and pager\-include\-volume (and similarly for the exclude variants).
.SH DISPLAY
The dockapp consists of a display window, two "action" buttons on the left, and
two "paging" buttons on the right. The paging buttons are used for paging
through the available devices; the mouse scrollwheel may also be used while
over the dockapp for the same purpose. The action buttons vary depending on
which device is focused.

Hovering the mouse over the display will show a tooltip with information on the
current device, and over the action buttons will show a tooltip describing the
button's action. Right-clicking the display gives a popup menu with additional
options for the current device. Middle-clicking moves back to FSM mode.

The "default" mode is the FSM display. In this mode, the filesystem usage
information for all filesystems reported as mounted by udisks is periodically
queried and the 6 with the highest percentage used are displayed, along with a
bar indicating the percentage. If the percentage is over 95%, the bar turns red
and (optionally) blinks. The action buttons are not used. If no
\fIclick-command\fP is set, clicking a filesystem line will go to that device's
page, if any.

The other mode is for displaying a device. In this mode, the top of the display
window shows the device label or /dev node name, with an icon representing the
device below. The right action button is used for removing the media as
applicable, e.g. locking a crypto device, ejecting a CD, or detaching a USB
stick for safe removal.

When displaying a filesystem device, the mountpoint and usage percent bar (if
mounted) are at the bottom of the display. The icon has a blue background if
the filesystem is mounted, or a red background if an operation is currently in
progress. The left action button is used to mount or unmount the filesystem.

When displaying a locked crypto device, the type of the device (e.g. "LUKS
device") is displayed at the bottom. The left action button is used to unlock
the device; the passphrase will be loaded from the system Secrets service if
possible, and you will be prompted if necessary.

The right-click menu gives access to a few additional actions, including
relabeling a filesystem and forgetting a saved crypto device password.
.SH BUGS
None yet.
.SH FILES
.TP
.I /etc/wmudmount.conf
System configuration.
.TP
.I $HOME/.wmudmountrc
User configuration.
.SH SIGNALS
.TP
.I SIGUSR1
Perform the same action as clicking the left action button.
.TP
.I SIGUSR2
Perform the same action as clicking the right action button.
.TP
.I SIGRTMIN+0
Perform the same action as clicking the left paging button. To find the expected signal number of SIGRTMIN+0, use the \fB--verbose\fP option.
.TP
.I SIGRTMIN+1
Perform the same action as clicking the right paging button. To find the expected signal number of SIGRTMIN+1, use the \fB--verbose\fP option.
.SH AUTHORS
\fBwmudmount\fP was written by \fIBrad Jorsch
<anomie@users.sourceforge.net>\fP.
.P
Email regarding wmudmount should be sent to
\fIanomie@users.sourceforge.net\fP.


install/TOOLS/dockapps/wmudmount/wmudmount.c
/* wmudmount
 * Copyright © 2010-2018  Brad Jorsch <anomie@users.sourceforge.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <fcntl.h>

#include <gtk/gtk.h>
#include <gdk/gdkx.h>

#include "askpass.h"
#include "conf.h"
#include "die.h"
#include "notify.h"
#include "chars.h"
#include "wmudisks.h"
#include "menu.h"
#include "dock.h"

/* Signal handling */
static int sig_fd[2];

static void sighandler(int sig){
    if (write(sig_fd[1], &sig, sizeof(sig)) < 0) {
        // Nothing much we can do about it inside a signal handler, but gcc
        // complains if we don't test it.
    }
    signal(sig, sighandler);
}

static gboolean sig_caught(GIOChannel *source G_GNUC_UNUSED, GIOCondition cond G_GNUC_UNUSED, gpointer data G_GNUC_UNUSED){
    int sig;
    while(read(sig_fd[0], &sig, sizeof(sig)) > 0){
        switch(sig){
          case SIGINT:
          case SIGQUIT:
          case SIGKILL: // Useless, but...
          case SIGPIPE:
          case SIGTERM:
          case SIGABRT:
            gtk_main_quit();
            break;
          case SIGHUP:
            // Reload something?
            break;
          case SIGUSR1:
            mount_button_click();
            redraw_dock();
            break;
          case SIGUSR2:
            eject_button_click();
            redraw_dock();
            break;
          default:
            if (sig == SIGRTMIN+0) {
                left_button_click();
                redraw_dock();
            } else if (sig == SIGRTMIN+1) {
                right_button_click();
                redraw_dock();
            }
            break;
        }
    }
    return TRUE;
}

/* Misc */

GQuark app_generic_error_quark(void) {
    return g_quark_from_static_string("app-generic-error-quark");
}

GQuark app_no_error_quark(void) {
    return g_quark_from_static_string("app-no-error-quark");
}

char *format_byte_size(guint64 size){
    static const char *byte_size_names[] = { "", "k", "M", "G", "T", "P", "E", "Z", "Y" };

    unsigned int i;
    double sz = size;
    for (i = 0; sz >= 1000 && i < sizeof(byte_size_names) - 1; i++, sz /= 1000.0);
    if (i == 0) {
        return g_strdup_printf("%ld", size);
    } else {
        return g_strdup_printf("%.2f %sB", sz, byte_size_names[i]);
    }
}

static gboolean parse_opt(const gchar *option_name, const gchar *value, gpointer data G_GNUC_UNUSED, GError **err){
    if(!strcmp(option_name,"-V") || !strcmp(option_name,"--version")){
        printf("%s\n", VERSION);
        exit(0);
    } else if(!strcmp(option_name,"-v") || !strcmp(option_name,"--verbose")){
        warn_level++;
    } else if(!strcmp(option_name,"-q") || !strcmp(option_name,"--quiet")){
        warn_level--;
    } else if(!strcmp(option_name,"-n") || !strcmp(option_name,"--notify")){
        if(!strcmp(value,"none")){
            notify_level = NOTIFY_NONE;
        } else if(!strcmp(value,"err") || !strcmp(value,"error")){
            notify_level = NOTIFY_ERROR;
        } else if(!strcmp(value,"warn") || !strcmp(value,"warning")){
            notify_level = NOTIFY_WARN;
        } else if(!strcmp(value,"info") || !strcmp(value,"all")){
            notify_level = NOTIFY_INFO;
        } else {
            g_set_error(err, G_OPTION_ERROR, G_OPTION_ERROR_FAILED, "Notification level '%s' is not recognized", value);
            return FALSE;
        }
    } else if(!strcmp(option_name,"--exclude-system-volumes") || !strcmp(option_name,"--skip-system-filesystems")){
        fsm_skip_root_only_volumes = TRUE;
        pager_skip_root_only_volumes = TRUE;
    } else if(!strcmp(option_name,"--include-system-volumes") || !strcmp(option_name,"--include-system-filesystems")){
        fsm_skip_root_only_volumes = FALSE;
        pager_skip_root_only_volumes = FALSE;
    } else if(!strcmp(option_name,"--fsm-exclude-system-volumes")){
        fsm_skip_root_only_volumes = TRUE;
    } else if(!strcmp(option_name,"--fsm-include-system-volumes")){
        fsm_skip_root_only_volumes = FALSE;
    } else if(!strcmp(option_name,"--pager-exclude-system-volumes")){
        pager_skip_root_only_volumes = TRUE;
    } else if(!strcmp(option_name,"--pager-include-system-volumes")){
        pager_skip_root_only_volumes = FALSE;
    }
    return TRUE;
}
static GOptionEntry entries[] = {
    {"version", 'V', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, parse_opt, "Print the version number and exit", NULL},
    {"no-default-config", 0, 0, G_OPTION_ARG_NONE, NULL, "Do not read the default configuration files", NULL},
    {"conf", 0, 0, G_OPTION_ARG_STRING, NULL, "Specify a configuration file to use in addition to the default", NULL},
    {"verbose", 'v', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, parse_opt, "Print more messages (can be repeated)", NULL},
    {"quiet", 'q', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, parse_opt, "Print fewer messages (can be repeated)", NULL},
#ifdef HAVE_LIBNOTIFY
    {"notify", 'n', 0, G_OPTION_ARG_CALLBACK, parse_opt, "Level of notifications to show: none, error, warn, info", NULL},
#endif
    {"return-to-fsm-timeout", 0, 0, G_OPTION_ARG_INT, &return_to_fsm_timeout, "Seconds before automatically returning to the FSM display (-1 to disable)", NULL},

    {"skip-system-filesystems", 0, G_OPTION_FLAG_HIDDEN|G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, parse_opt, NULL, NULL},
    {"include-system-filesystems", 0, G_OPTION_FLAG_HIDDEN|G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, parse_opt, NULL, NULL},
    {"exclude-system-volumes", 0, G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, parse_opt, "Do not include system volumes in the FSM or when paging", NULL},
    {"include-system-volumes", 0, G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, parse_opt, "Do not include system volumes in the FSM or when paging", NULL},
    {"fsm-exclude-system-volumes", 0, G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, parse_opt, "Do not include system volumes in the FSM", NULL},
    {"fsm-include-system-volumes", 0, G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, parse_opt, "Do not include system volumes in the FSM", NULL},
    {"pager-exclude-system-volumes", 0, G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, parse_opt, "Do not include system volumes when paging", NULL},
    {"pager-include-system-volumes", 0, G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, parse_opt, "Do not include system volumes when paging", NULL},

    {"allow-insecure-memory", 0, 0, G_OPTION_ARG_NONE, &allow_insecure_memory, "Allow passphrases to be temporarily stored in insecure memory", NULL},
    {"no-allow-insecure-memory", 0, G_OPTION_FLAG_REVERSE, G_OPTION_ARG_NONE, &allow_insecure_memory, "Exit if secure memory cannot be allocated", NULL},
    {"allow-core-files", 0, 0, G_OPTION_ARG_NONE, &allow_core_files, "Allow execution even if core file dumping cannot be disabled", NULL},
    {"no-allow-core-files", 0, G_OPTION_FLAG_REVERSE, G_OPTION_ARG_NONE, &allow_core_files, "Exit if core file dumping cannot be disabled", NULL},
    {"blink-full-filesystems", 0, 0, G_OPTION_ARG_NONE, &blink_full_filesystems, "Blink when a rw filesystem is over 95% full", NULL},
    {"no-blink-full-filesystems", 0, G_OPTION_FLAG_REVERSE, G_OPTION_ARG_NONE, &blink_full_filesystems, "Do not blink when a rw filesystem is over 95% full", NULL},
    {"non-wmaker", 0, 0, G_OPTION_ARG_NONE, &nonwmaker, "Use in a non-Window Maker window manager", NULL},
    {"use-secret-service", 0, 0, G_OPTION_ARG_NONE, &use_keyring, "Use the Secrets service for encrypted filesystem unlock passphrase storage", NULL},
    {"no-secret-service", 0, G_OPTION_FLAG_REVERSE, G_OPTION_ARG_NONE, &use_keyring, "Do not use the Secrets service for encrypted filesystem unlock passphrase storage", NULL},
    {NULL}
};

// This may not be defined anywhere, so define it now
int pipe2(int pipefd[2], int flags);

int main(int argc, char *argv[]){
    warn(DEBUG_INFO, "Parsing conf files");
    load_conf(&argc, argv);

    warn(DEBUG_INFO, "Initializing GTK");
    GOptionContext *context;
    context = g_option_context_new("");
    g_option_context_add_group(context, gtk_get_option_group(TRUE));
    g_option_context_add_main_entries(context, entries, NULL);
    GError *err = NULL;
    if(!g_option_context_parse(context, &argc, &argv, &err)){
        die("%s", err?err->message:"Failed to parse command line options");
    }
    g_clear_error(&err);
    g_option_context_free(context);

    init_notify();
    init_menu();
    create_dock_icon();

    warn(DEBUG_DEBUG, "Selecting and initializing Secrets service");
    init_keyring();

    warn(DEBUG_DEBUG, "Initializing udisks interface");
    if(!wmudisks_init()) die("Could not connect to the udisks daemon");

    warn(DEBUG_DEBUG, "Setting up signal handling");
    errno=0;
    if(!pipe2(sig_fd,O_NONBLOCK)){
        GIOChannel *channel = g_io_channel_unix_new(sig_fd[0]);
        g_io_add_watch(channel, G_IO_IN, sig_caught, NULL);
        g_io_channel_unref(channel);
        signal(SIGINT, sighandler);
        signal(SIGQUIT, sighandler);
        signal(SIGKILL, sighandler); // Useless, but...
        signal(SIGPIPE, sighandler);
        signal(SIGTERM, sighandler);
        signal(SIGABRT, sighandler);
        signal(SIGHUP, sighandler);
        signal(SIGUSR1, sighandler);
        signal(SIGUSR2, sighandler);
        signal(SIGRTMIN+0, sighandler);
        signal(SIGRTMIN+1, sighandler);
        warn(DEBUG_INFO, "SIGRTMIN+0 is %d", SIGRTMIN+0);
        warn(DEBUG_INFO, "SIGRTMIN+1 is %d", SIGRTMIN+1);
    } else {
        warn(DEBUG_WARN, "Signal handling initialization failed (%s), will probably die on any signal", strerror(errno));
    }

    warn(DEBUG_DEBUG, "Starting GTK main loop");
    gtk_main();
    warn(DEBUG_DEBUG, "GTK main loop exited");
    return 0;
}
